vllm/worker/worker_base.py

```diff
@@ -81,3 +87,54 @@ def remove_lora(self, lora_id: int) -> bool:
 
     def list_loras(self) -> List[int]:
         raise ValueError(f"{type(self)} does not support LoRA")
+
+
+class WorkerWrapperBase:
+    """
+    The whole point of this class is to lazily initialize the worker.
+    We first instantiate the WorkerWrapper, which remembers the worker module
+    and class name. Then, when we call `update_environment_variables`, and the
+    real initialization happens in `init_worker`.
+    """
+
+    def __init__(self,
+                 init_cached_hf_modules=False,
+                 worker_module_name=None,
+                 worker_class_name=None) -> None:
+        self.worker_module_name = worker_module_name
+        self.worker_class_name = worker_class_name
+        self.init_cached_hf_modules = init_cached_hf_modules
+        self.worker = None
+
+    def update_environment_variables(self, envs: Dict[str, str]) -> None:
+        key = 'CUDA_VISIBLE_DEVICES'
+        if key in envs and key in os.environ:
+            # overwriting CUDA_VISIBLE_DEVICES is desired behavior
+            # suppress the warning in `update_environment_variables`
+            del os.environ[key]
+        update_environment_variables(envs)
```

Comment:
At a high-level, we should discourage passing configuration / data via environment variables because it allows for a lot of complexity that isn't subject to documentation / visibility / common abstractions that a config object would be subject to. Basically, we don't want to give the ability for external contributors to add additional configuration space very easily. An example of this is the Ray DAG vLLM integration, which is very hidden because it's not in a config object. https://github.com/vllm-project/vllm/blob/e95cd879598b834f85e70ebcd23db316ae430540/vllm/executor/ray_gpu_executor.py#L24-L27

Another example is that `CUDA_VISIBLE_DEVICES` semantics may only work for a subset of backends; the way each backend configures which devices to use can be different. Then we have a cambrian explosion of env vars to manage devices..

Alternatives:
* Allow a whitelist of env vars that we're confident apply for all vLLM backends, e.g. MPI's `RANK`/`LOCAL_RANK`/`WORLD_SIZE`/etc
* Take standard vLLM config and allow each backend to interpret / apply it (e.g. `set_visible_devices` API instead of `update_environment_variables`)

Reply:
Agree! and I won't block the PR on it -- but to raise vLLM quality as we have more backends, we should be careful with global state https://softwareengineering.stackexchange.com/questions/148108/why-is-global-state-so-evil

OK to have a small set of allowed env vars, but it's a bad idea to support arbitrary ones at interface level. fine to improve later.